### PR TITLE
az://303 vclock enhancements redux be2

### PR DESCRIPTION
**Pull request by jonmeredith**
az303 redux branch rebased against be phase 2, use in conjunction with the same named riak_core branch.

To get the new vclock behavior you must set {vnode_vclocks, true} before retrieving riak_client objects.

This should address final comments about allow_mult in https://github.com/basho/riak_kv/pull/157.

Bryan requested whoever reviewed double-checked deletes.

### CODE REVIEW

File: src/riak_kv_put_fsm.erl [hunk 1]
```diff
@@ -162,23 +167,42 @@ init({test, Args, StateProps}) ->
...
-    new_state_timeout(validate, StateData).
-    
+    %% Check if this node is in the preference list so it can coordinate
+    case ([IndexNode || {{_Index, Node} = IndexNode, _Type} <- Preflist2,
+                        Node == node()]) of
+        [] ->
+            %% This node is not in the preference list
+            %% forward on to the first node
+            [{{_Idx, CoordNode},_Type}|_] = Preflist2,
+            %% io:format(user, "Forwarding ~p to coordinator on ~p\n", [BKey, CoordNode]),
```
Comment (kellymclaughlin) [inline src/riak_kv_put_fsm.erl]: Get rid of debug output.

File: src/riak_kv_put_fsm.erl [hunk 2]
```diff
@@ -382,6 +428,27 @@ process_reply(Reply, StateData = #state{postcommit = PostCommit,
             new_state_timeout(finish, StateData2)
     end.
 
+
+%% @private
+%% Send the put requests to any remote nodes if necessary and decided if 
+%% enough responses have been received yet (i.e. if W/DW=1)
+execute_remote(StateData=#state{robj=RObj, req_id = ReqId,
```
Comment (kellymclaughlin) [inline src/riak_kv_put_fsm.erl:436]: What about moving `execute_remote` before `waiting_remote_vnode`. That way the states would be in the order of their normal transition. Doesn't matter functionally, but makes more sense for reading the code.

File: test/ec_eqc.erl [hunk 3]
```diff
@@ -0,0 +1,955 @@
...
+test(N) ->
+    quickcheck(numtests(N, prop())).
+
+check() ->
+    check(prop(), current_counterexample()).
+
+-define(B, <<"b">>).
+-define(K, <<"k">>).
+
+                                                %-define(FINALDBG(Fmt,Args), io:format("= " ++ Fmt, Args)).
```
Comment (kellymclaughlin) [inline test/ec_eqc.erl]: Add an extra `%` here so the emacs formatting doesn't keep sending him way over there.

File: test/ec_eqc.erl [hunk 4]
```diff
@@ -0,0 +1,955 @@
...
+gen_client_seeds() ->
+    ?LET({Clients, Handoff}, 
+         {non_empty(list(gen_regular_client())), gen_handoff_client()},
+         [Handoff | Clients]).
+
+prop() ->
+    ?FORALL({Pris, ClientSeeds, ParamsSeed},
+            {gen_pris(), gen_client_seeds(),gen_params()},
+            begin
+                %% io:format(user, "Pris: ~p\n", [Pris]),
```
Comment (kellymclaughlin) [inline test/ec_eqc.erl:123]: Not sure if this was left intentionally or not so just point it out in case you meant to remove.

**Comment by kellymclaughlin** [comment]
Seeing a problem with indexes missing after an update with `allow_mult=false`. Steps to reproduce using index backend:
- curl -v -X PUT -H "Content-Type: application/json" \
  -d '{"props":{"allow_mult":true}}' \
  http://127.0.0.1:8098/riak/kitchen
- curl -v -X POST -H "Content-Type: application/json" \
  -H "x-riak-index-field1_bin: val1" \
  -H "x-riak-index-field2_int: 1001" \
  -d '{"dishes":11}' \
  http://127.0.0.1:8098/riak/kitchen/sink?returnbody=true
- curl -v -X PUT -H "Content-Type: application/json" \
  -H "x-riak-index-field1_bin: val2" \
  -H "x-riak-index-field2_int: 1003" \
  -d '{"dishes":9}' \
  http://127.0.0.1:8098/riak/kitchen/sink?returnbody=true

Then `curl -v localhost:8098/buckets/kitchen/index/field1_bin/eq/val2` should return `sink`, but I get an empty list.

File: src/riak_kv_put_fsm.erl (hunk 1 above)
Comment (jonmeredith) [inline src/riak_kv_put_fsm.erl]: Removed - added error log for forwarding failure.

File: src/riak_kv_put_fsm.erl (hunk 2 above)
Comment (jonmeredith) [inline src/riak_kv_put_fsm.erl:436]: I need to add a variant of the put FSM that just stores and does not increment for repl - I'll restructure then.  It does read better that way but it makes it seem like execute_remote is a state when it isn't.

File: test/ec_eqc.erl (hunk 3 above)
Comment (jonmeredith) [inline test/ec_eqc.erl]: Fixed.

**Comment by jonmeredith** [comment]
I made a mistake in the refactor to combine to the put paths - for a coordinated put the vnode has to return the object.  With it in place I get this result.  I think the same issue was affecting the keys_fsm_eqc test.

```
jons-macpro:riak_kv(az303-vclock-enhancements-redux-be2) jmeredith$ curl -v localhost:8091/buckets/kitchen/index/field1_bin/eq/val2
* About to connect() to localhost port 8091 (#0)
*   Trying 127.0.0.1... connected
* Connected to localhost (127.0.0.1) port 8091 (#0)
> GET /buckets/kitchen/index/field1_bin/eq/val2 HTTP/1.1
> User-Agent: curl/7.19.7 (universal-apple-darwin10.0) libcurl/7.19.7 OpenSSL/0.9.8r zlib/1.2.3
> Host: localhost:8091
> Accept: */*
> 
< HTTP/1.1 200 OK
< Vary: Accept-Encoding
< Server: MochiWeb/1.1 WebMachine/1.8.1 (participate in the frantic)
< Date: Sat, 03 Sep 2011 21:43:34 GMT
< Content-Type: application/json
< Content-Length: 17
< 
* Connection #0 to host localhost left intact
* Closing connection #0
{"keys":["sink"]}
```

File: src/riak_kv_put_fsm.erl (hunk 2 above)
Comment (kellymclaughlin) [inline src/riak_kv_put_fsm.erl:436]: Nevermind, I didn't pay close enough attention and thought it was a state.

**Comment by kellymclaughlin** [comment]
The keys_fsm eqc test is working consistently for me now and the index issue is resolved. I'll probably just make one more pass through the code before giving it the green light for merge.

**Comment by jonmeredith** [comment]
Replaced by https://github.com/basho/riak_kv/pull/173 which is rebased off master post-cluster membership merge.

**Pull request closed without merging**